### PR TITLE
BUG: Use whole file for encoding checks with ``charset_normalizer``.

### LINKED ISSUE #22871: BUG: f2py cannot handle non-ascii characters in comments since 1.24
### Describe the issue:

F2py works fine in version 1.23 when an utf8-encoded Fortran file has non-ascii characters in comments, but gives an error in 1.24. The error message recommends installing `chardet` but it does not seem to fix the issue.

### Reproduce the code example:

```
$ cat test.f90 
subroutine foo(x)
  real(8), intent(in) :: x
  print*, x ! Écrit à l'écran la valeur de x
end subroutine

$ conda create -n numpy1.23 numpy=1.23
[...]
$ conda create -n numpy1.24 numpy=1.24
[...]
$ conda activate numpy1.23
$ f2py -m test test.f90
Reading fortran codes...
	Reading file 'test.f90' (format:free)
Post-processing...
	Block: test
			Block: foo
Post-processing (stage 2)...
Building modules...
    Building module "test"...
    Generating possibly empty wrappers"
    Maybe empty "test-f2pywrappers.f"
        Constructing wrapper function "foo"...
          foo(x)
    Wrote C/API module "test" to file "./testmodule.c"
$ conda activate numpy1.24
$ f2py -m test test.f90
Reading fortran codes...
Traceback (most recent call last):
[...]
```

### Error message:

```
Traceback (most recent call last):
  File "/opt/mancellin/mambaforge/envs/numpy1.24/lib/python3.11/site-packages/numpy/f2py/crackfortran.py", line 393, in readfortrancode
    l = fin.readline()
        ^^^^^^^^^^^^^^
  File "/opt/mancellin/mambaforge/envs/numpy1.24/lib/python3.11/fileinput.py", line 292, in readline
    line = self._readline()
           ^^^^^^^^^^^^^^^^
  File "/opt/mancellin/mambaforge/envs/numpy1.24/lib/python3.11/fileinput.py", line 372, in _readline
    return self._readline()
           ^^^^^^^^^^^^^^^^
  File "/opt/mancellin/mambaforge/envs/numpy1.24/lib/python3.11/encodings/ascii.py", line 26, in decode
    return codecs.ascii_decode(input, self.errors)[0]
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
UnicodeDecodeError: 'ascii' codec can't decode byte 0xc3 in position 59: ordinal not in range(128)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/opt/mancellin/mambaforge/envs/numpy1.24/bin/f2py", line 10, in <module>
    sys.exit(main())
             ^^^^^^
  File "/opt/mancellin/mambaforge/envs/numpy1.24/lib/python3.11/site-packages/numpy/f2py/f2py2e.py", line 704, in main
    run_main(sys.argv[1:])
  File "/opt/mancellin/mambaforge/envs/numpy1.24/lib/python3.11/site-packages/numpy/f2py/f2py2e.py", line 441, in run_main
    postlist = callcrackfortran(files, options)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/mancellin/mambaforge/envs/numpy1.24/lib/python3.11/site-packages/numpy/f2py/f2py2e.py", line 342, in callcrackfortran
    postlist = crackfortran.crackfortran(files)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/mancellin/mambaforge/envs/numpy1.24/lib/python3.11/site-packages/numpy/f2py/crackfortran.py", line 3298, in crackfortran
    readfortrancode(files, crackline)
  File "/opt/mancellin/mambaforge/envs/numpy1.24/lib/python3.11/site-packages/numpy/f2py/crackfortran.py", line 395, in readfortrancode
    raise Exception(
Exception: readfortrancode: reading test.f90#0 failed with
'ascii' codec can't decode byte 0xc3 in position 59: ordinal not in range(128).
It is likely that installing chardet package will help f2py determine the input file encoding correctly.
```

### Runtime information:

>>> import sys, numpy; print(numpy.__version__); print(sys.version)
1.24.0
3.11.0 | packaged by conda-forge | (main, Oct 25 2022, 06:24:40) [GCC 10.4.0]

### Context for the issue:

I can't build the Fortran dependency of my package with Numpy 1.24. See also https://github.com/capytaine/capytaine/issues/273

### PR DESCRIPTION
**Pull request by HaoZeke**
Closes #22871.

The issue is that the current behavior uses only the first `32` bytes. The current variation (reading in the whole file) should be fine, fortran files are rarely large enough for this to be a practical bottleneck (tentatively).

**EDIT**: Now this has a slightly larger changelog

- [x] Switch from `chardet` to `charset_normalizer`
- [x] Rework to keep old behavior (bytes read) without `charset_normalizer`

The rationale here is that if there are encoding errors, attempting to determine the encoding with `startswith` doesn't need more than the old number of bytes anyway.

`charset_normalizer` will use the whole file to check the encoding.

### CODE REVIEW

**Comment by charris** [comment]
@HaoZeke Could you add a test for this?

The travis failures can be ignored, the cause is known and will be fixed soon.

**Comment by HaoZeke** [comment]
> @HaoZeke Could you add a test for this?

Sure, in a bit, I'll have to add `chardet` to the CI though, since it is needed to pass the test. Actually, maybe it should be an optional requirement of `numpy` now?

> The travis failures can be ignored, the cause is known and will be fixed soon.

Excellent.

**Comment by melissawm** [comment]
> fortran files are rarely large enough for this to be a practical bottleneck (tentatively).

As a nitpick, should we also add this as a comment either on the code or the test you are going to add? Other than that, looks fine to me :)

**Comment by charris** [comment]
Hmm. Looks like `chardet` is under the LGPL. It is already an optional dependency of `crackfortran.py`, so I guess making it an optional dependency for testing doesn't change anything, but I do wonder if we have any related license requirements because it is in the code, i.e., mentioning the license, even when `charded` use is optional. IANAL and have trouble understanding the license text, but since use is optional perhaps the requirements don't apply. @rgommers Thoughts?

**Comment by charris** [comment]
@HaoZeke If we keep the dependency, it should also be added to `test_requirements.txt`.

**Comment by mattip** [comment]
Maybe we can switch from chardet to the MIT-licensed [charset_normalizer](https://charset-normalizer.readthedocs.io/en/latest/), there is a [FAQ about compatibility](https://charset-normalizer.readthedocs.io/en/latest/community/faq.html#is-it-backward-compatible-with-chardet) with chardet

**Comment by HaoZeke** [comment]
`charset_normalizer` seems to be better maintained and works just as well. Also there's a convenience function `from_path` which cleans up the original logic a bit too, so it seems like a win-win.

**Comment by HaoZeke** [comment]
Some of the newest change disallows previously allowed (but highly unlikely to be present) code-paths, namely having fixed form F77 code in a fortran 90 file (with `.f90`).

Does this maybe need a release note now?

**Comment by charris** [comment]
> EDIT: I'm moving this change out of this PR to #22885, the tests can be fixed without it.

Thanks, I was going to ask for that.

File: numpy/f2py/tests/test_crackfortran.py [hunk 1]
```diff
@@ -258,12 +260,24 @@ def test_input_encoding(self, tmp_path, encoding):
         # gh-635
         f_path = tmp_path / f"input_with_{encoding}_encoding.f90"
         # explicit BOM is required for UTF8
-        bom = {'utf-8': codecs.BOM_UTF8}.get(encoding, b'')
+        if "ecoding" == 'utf-8':
```
Comment (charris) [inline numpy/f2py/tests/test_crackfortran.py:263]: Spelling?

Comment (HaoZeke) [reply numpy/f2py/tests/test_crackfortran.py]: Sorry, excellent catch, this doesn't seem to be required anymore at all, simplifying.

**Review by charris — APPROVED** [review]

**Comment by charris** [comment]
Thanks @HaoZeke .

**Pull request merged**